### PR TITLE
Add new sequential and multicore versions of mergesort 

### PR DESCRIPTION
**Pull request by moazzammoriani**
This PR addresses #356.

### CODE REVIEW

File: benchmarks/multicore-numerical/mergesort.ml [hunk 1]
```diff
@@ -1,42 +1,62 @@
...
+        (Array.sort (fun x y -> x - y) c) in 
+    print_endline ((c = a) |>  Bool.to_string);
+    print_endline (string_of_int n);
+
+
+
+(*let () =
+  let pool = T.setup_pool ~num_additional_domains:(num_domains - 1) () in
+  sort pool a;
+  T.teardown_pool pool *)
```
Comment (Sudha247) [inline benchmarks/multicore-numerical/mergesort.ml:62]: The commented code can be removed.

**Review by Sudha247 — COMMENTED** [review]
Thanks for the PR, @moazzammoriani! The changes look good to me. Are you able to get the running times on Godel please? It's also worth revisiting the run configs and adjust the input params if needed.

**Comment by moazzammoriani** [comment]
The sequential version of mergesort takes 1m24s to run on the gödel machine for an input of size 2^26. To my surprise, while running the multicore version, the allocation of a single domain was enough to significantly lower the runtime to 26s. The rest of the results, as a function of the number of cores for this input size, on the gödel machine, are listed below. 

```
++ exe=_build/default/mergesort_multicore.exe
++ arg=67108864
++ for i in 1 2 4 8 12
++ taskset --cpu-list 2-13 chrt -r 1 _build/default/mergesort_multicore.exe 1 67108864

real	0m26.404s
user	0m25.864s
sys	0m0.540s
++ for i in 1 2 4 8 12
++ taskset --cpu-list 2-13 chrt -r 1 _build/default/mergesort_multicore.exe 2 67108864

real	0m15.639s
user	0m26.374s
sys	0m0.552s
++ for i in 1 2 4 8 12
++ taskset --cpu-list 2-13 chrt -r 1 _build/default/mergesort_multicore.exe 4 67108864

real	0m10.305s
user	0m26.855s
sys	0m0.616s
++ for i in 1 2 4 8 12
++ taskset --cpu-list 2-13 chrt -r 1 _build/default/mergesort_multicore.exe 8 67108864

real	0m7.843s
user	0m27.497s
sys	0m0.469s
++ for i in 1 2 4 8 12
++ taskset --cpu-list 2-13 chrt -r 1 _build/default/mergesort_multicore.exe 12 67108864

real	0m6.879s
user	0m27.761s
sys	0m0.525s
++ for i in 16 20 24
++ taskset --cpu-list 2-13,16-27 chrt -r 1 _build/default/mergesort_multicore.exe 16 67108864

real	0m6.732s
user	0m33.190s
sys	0m0.549s
++ for i in 16 20 24
++ taskset --cpu-list 2-13,16-27 chrt -r 1 _build/default/mergesort_multicore.exe 20 67108864

real	0m6.825s
user	0m40.724s
sys	0m0.541s
++ for i in 16 20 24
++ taskset --cpu-list 2-13,16-27 chrt -r 1 _build/default/mergesort_multicore.exe 24 67108864

real	0m6.611s
user	0m45.517s
sys	0m0.590s
```

**Review by Sudha247 — COMMENTED** [review]
The code looks good to me. I think we know why there's a discrepancy in time between the sequential and parallel versions. Could you please add entries for this in `run_config.json` and `multicore_parallel_*` config files. The PR can then be merged.

File: benchmarks/multicore-numerical/mergesort.ml [hunk 2]
```diff
@@ -1,42 +1,55 @@
...
+let sort a =
+  let b = Array.copy a in
+   merge_sort false a b 0 (Array.length a)
+
+let () = 
+    let c = Array.copy a in 
+    let () =  
+        (sort a) in 
+    let () = 
+        (Array.sort (fun x y -> x - y) c) in
```
[the remark this answers is not in the document]
Comment (moazzammoriani) [reply benchmarks/multicore-numerical/mergesort.ml:53]: Ah that was perhaps a bit careless. I'll fix it immediately.

File: run_config.json [hunk 3]
```diff
@@ -289,7 +289,18 @@
         }
       ]
     },
-
+    {
+      "executable": "benchmarks/multicore-numerical/nqueens.exe",
```
Comment (Sudha247) [inline run_config.json:293]:
```suggestion
      "executable": "benchmarks/multicore-numerical/mergesort.exe",
```

**Review by Sudha247 — APPROVED** [review]
Minor typo, looks good otherwise.

**Comment by moazzammoriani** [comment]
@Sudha247 should be good now I think.

**Comment by Sudha247** [comment]
Thanks!

**Pull request merged**